main/tests/server/src/com/google/refine/RefineTest.java

```diff
@@ -431,6 +434,29 @@ protected ButterflyModule getCoreModule() {
         return coreModule;
     }
 
+    /**
+     * Runs an operation on a project. If it's a long-running operation, its process is run in the main thread until
+     * completion.
+     * 
+     * @returns the duration of the operation in milliseconds
+     */
+    protected long runOperation(AbstractOperation operation, Project project) throws Exception {
+        long start = System.currentTimeMillis();
+        Process process = operation.createProcess(project, new Properties());
+        if (process.isImmediate()) {
+            process.performImmediate();
+        } else {
+            LongRunningProcessStub longRunning = new LongRunningProcessStub((LongRunningProcess) process);
+            longRunning.run();
```

Comment:
Because this is running on the main thread an unhandled exception should propagate, what happens if this runs forever (or even above some upper bound)?

If we don't have one already, it might be worth adding a test to make sure that exception handling works the way we expect. Also some separate explicit tests of the long running process lifecycle (isRunning, etc).

Reply:
If the operation runs forever, then the test will run forever indeed. That's a downside of this approach.
For me, the main motivation is to reduce the resource usage of the test suite, by avoiding the creation of many threads (which are not even reused via a thread pool, as is common practice), and simplify debugging.

If you prefer keeping it in a separate thread, I'd be happy revert back to that. My main motivation here is to introduce this test utility to simplify tests and minimize the changes I need to do later on. The specifics of how operations are executed is not so important to me at this stage because that will change completely with the new architecture anyway.

Reply:
Let's put it back to using a separate thread then, if you're OK with that. It keeps things closer aligned with the production code and we're talking hundreds or perhaps thousands of threads for the test suite, not millions.

My rough timings indicate that the savings is on the order of 5% and dropping the polling interval would get some of that back.